src/main/scala/histogramactors/HistogramMain.scala

```diff
@@ -0,0 +1,137 @@
+package histogramactors
+
+import akka.actor.{Actor, ActorRef, ActorSystem, Props}
+import akka.pattern.{ask, pipe}
+import akka.util.Timeout
+
+import java.io.PrintWriter
+import scala.collection.mutable
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.concurrent.duration.{Duration, MILLISECONDS}
+import scala.io.Source
+import scala.util.{Failure, Success, Try, Using}
+
+object HistogramMain extends App {
+  val inputFilePath = "src/main/resources/histogram/input.txt"
+  val outputFilePath = "src/main/resources/histogram/output.txt"
+
+  val system = ActorSystem("histogram")
+  val master = system.actorOf(Props(classOf[Master], inputFilePath, outputFilePath, system), "master")
+  master ! Start
+
+}
+
+class Master(input: String, output: String, system: ActorSystem) extends Actor {
+  private val poolSize = 10
+
+  val loader: ActorRef = system.actorOf(Props[Loader](), "loader")
+  val reducer: ActorRef = system.actorOf(Props[Reducer], "reducer")
+  val parserPool: ActorRef = system.actorOf(Props(classOf[ParserPool], poolSize), "parserPool")
+  val mapper: ActorRef = system.actorOf(Props(classOf[Mapper], parserPool, reducer), "mapper")
+  val saver: ActorRef = system.actorOf(Props[Saver](), "saver")
+  override def receive: Receive = {
+    case Start =>
+      loader ! Load(input)
+    case LoadResult(dataFromFile) =>
+      mapper ! DocumentToMap(dataFromFile)
+    case MappingFinished =>
+      reducer ! GetReduceResult
+    case ReduceResult(result) =>
+      saver ! Save(output, result)
+    case FailResult(ex) =>
+      throw ex
+      system.terminate()
+    case SuccessResult =>
+      system.terminate()
+  }
+}
+
+class Loader extends Actor {
+  override def receive: Receive = {
+    case Load(filePath) =>
+      loadData(filePath) match {
+        case Success(dataFromFile) => sender() ! LoadResult(dataFromFile)
```

Comment:
Bad practice calling blocking operations in receive, try using a future.

Reply:
Call of loadData function has been added to Future